File: python/4_mocked_logic_and_testcases_test.py
import os
import pytest
from unittest import mock


def get_filenames_in_directory(directory):
    fileNames = []
    for root, dirs, files in os.walk(directory):
        for name in files:
            fileNames.append(name)
    return fileNames


# https://docs.pytest.org/en/latest/parametrize.html
# https://docs.python.org/3/library/unittest.mock.html
@mock.patch('os.walk')
@pytest.mark.parametrize('files,expected_count', [
  (['a.txt', 'b.txt', 'c.png'], 3),
  (['a.txt', 'c.png'], 2),
  ([], 0),
  (['a', 'b', 'c', 'd', 'e', 'f', 'g', 'h', 'i', 'j'], 10)
])
def test_with_mocked_logic_and_test_cases(mockwalk, files, expected_count):
    folders = ['folder1', 'folder2']
    mockwalk.return_value = [('/root', folders, files)]

    result = get_filenames_in_directory('path')
    assert len(result) == expected_count
    assert result == files
    mockwalk.assert_called_once_with('path')

File: js/4_mockedLogicAndTestcasesTest.js
const fs = require('fs');
const sinon = require('sinon');
const assert = require('assert');

function getFilenamesInDirectory(directory) {
    let fileNames = [];
    fs.readdir(directory, (err, files) => {
        files.forEach(file => {
            fileNames.push(file);
        });
    });
    return fileNames;
}

// https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Array/forEach
// http://sinonjs.org/releases/v3.3.0/stubs/
describe('Test that testcases work in parallel with stubbed logic', function () {
    const tests = [
        {files: ['a.txt', 'b.txt', 'c.png']},
        {files: ['a.txt', 'c.png']},
        {files: []},
        {files: ['a', 'b', 'c', 'd', 'e', 'f', 'g', 'h', 'i', 'j']}
    ];

    tests.forEach(function(test) {
        it('stub should return list of files containing ' + test.files, function () {
            let fsStub = sinon.stub(fs, 'readdir').callsFake(function(path, callback) {
                callback(null, test.files);
            });

            const result = getFilenamesInDirectory('path');
            assert.equal(result.length, test.files.length);
            assert.deepEqual(result, test.files); // structual equality (checks that operands are equivalent)

            fsStub.restore();
        });
    });
});
